### PR TITLE
Create reusable fakes/stubs for VSCode UI components

### PR DESCRIPTION
**Pull request by ckant**
## Description

Create reusable fakes/stubs for VSCode UI components

## Motivation and Context

These components are meant to be injected and used in place of the functions
in the vscode namespace.

Default implementations are provided that simply delegate to vscode.

Fake implementations capture and/or control the outputs
to verify hard-to-test situations without needing separate UI tests.

e.g.

- Verifying progress bars progression
- Verifying messaging on input box validation failures
- Faking commands that would update the UI

They also allow for capturing events that modify global state,
such as updating the system clipboard.

This implementation only covers a few portions of the vscode APIs.
It can be incrementally built out as functionality is needed.

### Used in the following PRs

* [Add S3 nodes to AWS Explorer tree](https://github.com/aws/aws-toolkit-vscode/pull/1079) #1079
* [Introduce commands for acting on S3 nodes / resources ](https://github.com/aws/aws-toolkit-vscode/pull/1080) #1080

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the changelog using the script `npm run newChange`

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/vscode/window.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+/**
+ * Actions associated with showing messages in {@link vscode.window}.
+ */
+export interface Message {
+    /**
+     * See {@link vscode.window.showErrorMessage}.
+     */
+    showError(message: string, ...items: string[]): Thenable<string | undefined>
```
Comment (awschristou) [inline src/shared/vscode/window.ts:46]: You might as well add showInformation and showWarning from the start.

File: src/test/shared/vscode/fakeEnv.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+export class FakeEnv implements Env {
+    private readonly _clipboard: FakeClipboard
+
+    public get clipboard(): FakeClipboard {
+        return this._clipboard
+    }
+
+    public constructor({ clipboard = new FakeClipboard() }: FakeEnvOptions = {}) {
+        this._clipboard = clipboard instanceof FakeClipboard ? clipboard : new FakeClipboard(clipboard)
```
Comment (awschristou) [inline src/test/shared/vscode/fakeEnv.ts:20]: if clipboard isn't what you expected it to be, consider throwing instead of masking the issue with a different object, because the test was created incorrectly.

This comment applies to fakeWindow, and anywhere else that does this.

File: src/shared/vscode/env.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// @ts-ignore
+import * as vscode from 'vscode'
+
+/**
+ * Components associated with {@link vscode.env}.
+ */
+export interface Env {
```
Comment (awschristou) [inline src/shared/vscode/env.ts:12]: Do you envision objects like this getting passed around to every function in many call chains? Curious to understand the usage.

File: src/shared/vscode/window.ts (hunk 1 above)
Comment (awschristou) [inline src/shared/vscode/window.ts:46]: ShowXxxMessage are good examples in the vscode api where there are a bunch of overloads (see vscode's index.d.ts file). These are gross to work with at the mocking level. Do you have a plan to minimize pain in this space?

(an example use-case is to call a modal ShowXXXMessage )

File: src/shared/vscode/window.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    inputBox: InputBox
+    message: Message
+    progress: Progress
+    dialog: Dialog
+}
+
+/**
+ * Actions associated with status bars in {@link vscode.window}.
+ */
+export interface StatusBar {
```
Comment (awschristou) [inline src/shared/vscode/window.ts:22]: Consider supporting the equivalent of `createStatusBarItem`, since we aren't bound to just one status bar.

File: src/shared/vscode/defaultWindow.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        return vscode.window.withProgress(options, task)
+    }
+}
+
+export class DefaultDialog implements Dialog {
+    public showOpen(options: vscode.OpenDialogOptions): Thenable<vscode.Uri[] | undefined> {
+        return vscode.window.showOpenDialog(options)
+    }
+
+    public showSave(options: vscode.SaveDialogOptions): Thenable<vscode.Uri | undefined> {
```
Comment (bryceitoc9) [inline src/shared/vscode/defaultWindow.ts:54]: Looking at how this is implemented in https://github.com/aws/aws-toolkit-vscode/pull/1080 , I'm thinking I'd rather see this as mapping directly to the VS Code schema. Instead of `window.Dialog.showSave`, I'd rather see `window.showSaveDialog`. 

Goes the same for everything else here.

File: src/shared/vscode/window.ts (hunk 1 above)
Comment (ckant) [reply src/shared/vscode/window.ts:46]: Yeah, I was just adding the stuff I needed for the S3 integration. Will add those two as well.

File: src/test/shared/vscode/fakeEnv.ts (hunk 2 above)
Comment (ckant) [reply src/test/shared/vscode/fakeEnv.ts:20]: This is an attempt to implement a pattern where the user can pass either an instance of the class OR just the constructor args.

This reduces the verbosity with nested faking.

```ts
new FakeWindow({ 
  inputBox: { input: bucketName },
  dialog: { saveSelection: saveLocation },
})
```

versus
```ts
new FakeWindow({ 
  inputBox: new FakeInputBox({ input: bucketName }),
  dialog: new FakeDialog({ saveSelection: saveLocation }),
})
```

File: src/shared/vscode/env.ts (hunk 3 above)
Comment (ckant) [reply src/shared/vscode/env.ts:12]: I expect it to be injected into the class/function that needs it. So the caller shouldn't need to pass it around.

I've been using it as a default parameter on (command) class constructors. [(example)](https://github.com/aws/aws-toolkit-vscode/blob/kantc/4commands/src/awsexplorer/commands/copyName.ts#L17).

```ts
export async function copyNameCommand(
    node: AWSResourceNode,
    window: Window = new DefaultWindow(),
    env: Env = new DefaultEnv()
): Promise<void> {
```

The Window and Env fakes would be injected for unit testing. This doesn't solve the problem of trying to test a class which uses a class which uses this. We'd need real dependency injection or module replacement to do that. 

However, I've found that unit tests are quite sufficient for testing the kind of edge cases and strict validation that this is intended to help verify. Integration tests can usually use the real implementations and perform less precise validations.

File: src/shared/vscode/window.ts (hunk 1 above)
Comment (ckant) [reply src/shared/vscode/window.ts:46]: There are indeed a lot of overloads and variations of a lot of these. Luckily I think we're only using a small subset of them for normal cases. I think we can implement them as they come up. 

Branches in the logic caused by different parameters do make the mocking more difficult. However, most of these implementations simply capture the arguments and return a stub response. So dealing with variations in parameters shouldn't be nearly as difficult as in the implementation of the real vscode apis.

File: src/shared/vscode/defaultWindow.ts (hunk 5 above)
Comment (ckant) [reply src/shared/vscode/defaultWindow.ts:54]: Yeah I considered how much to stray from the original lumping of functions into namespaces. The difficulty with that approach is in discovering the inputs and outputs you can get/set. The namespaces are very large. Additionally, splitting them up into separate classes makes it so you can inject only the subset of window components you're using.

How would you suggest setting the user input for a save dialog?

Right now it's 

`window.dialog.saveSelection = 'blah.jpg'`

My thought was that this would be harder to discover among the many other properties attached to `window`.

`window.saveSelection = 'blah.jpg'`

Other options:
* `window.showSaveDialog(...)` is the function, but you use `window.dialogOptions.saveSelection` to get and set properties of it
* Attach the properties directly to the functions
  * `window.showSaveDialog.saveSelection`
* Support both call patterns

**Comment by ckant** [comment]
> I'm cautious on this -- interested to see what it looks like in use, but also understanding
> 
>     * how we keep from getting stung by interface changes when we update the engine version
> 
>     * what the cost is to increase increase the surface area to account for additional vs code api usage

My thought is that this should be no different to our existing code. If backwards incompatible changes are introduced into the vscode api, we'll break in either case. If new functionality is introduced, we'll have to change your code to use it in either case.

Around the codebase, we tend to resort to this same approach whenever we want to mock something for tests, albeit we make custom stubs each time and duplicate effort there. The goal is to unify these for injection.

We could also do the common JS approach of replacing the module implementation (using sinon/jest/etc). However the implementation there is necessarily very hacky. Additionally, dependency injection leads to better design in the long run as it makes the dev think about enforcing single responsibility and coupling when the dependencies are out of hand.

I think dependency injection is growing in popularity in JS/TS and will likely really take off once [ES gets official decorator support](https://github.com/tc39/proposal-decorators).

https://github.com/microsoft/tsyringe
http://inversify.io/
https://github.com/thiagobustamante/typescript-ioc

File: src/test/utilities/mockito.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+// Workaround https://github.com/NagRock/ts-mockito/issues/163
+export function mock<T>(): T {
+    const mocker = new Mocker(undefined)
+    mocker['excludedPropertyNames'] += 'then'
+
+    return mocker.getMock()
+}
+
+export const spy = mockito.spy
```
Comment (justinmk3) [inline src/test/utilities/mockito.ts:17]: why abstract this?

File: src/test/shared/vscode/fakeEnv.ts (hunk 2 above)
Comment (awschristou) [reply src/test/shared/vscode/fakeEnv.ts:20]: Ok I follow that, thanks for explaining.

File: src/test/utilities/mockito.ts (hunk 6 above)
Comment (ckant) [reply src/test/utilities/mockito.ts:17]: Thanks, changed to `export {spy} from 'ts-mockito'`

File: src/shared/vscode/defaultWindow.ts (hunk 5 above)
Comment (ckant) [reply src/shared/vscode/defaultWindow.ts:54]: Currently going with 

* Function call is `window.showSaveDialog(...)`
* Inputs are set up via `new FakeWindow(dialog: { saveSelection: 'foo' })`
* Outputs are retrieved via `window.dialog.{ saveOptions }`
